### PR TITLE
parse samples, run on one or all samples

### LINKED ISSUE #16: Parsing and Running on Samples for USACO Problems

### PR DESCRIPTION
**Pull request by bqi343**
at some point, should modify the callback for the run code button such that it runs the currently selected input tab. 
Closes #16 

i.e.,

- [x]  if you have a sample tab open, you will run that sample
- [x] if you have the judge tab open, you will run all samples

store verdicts and outputs separately for every sample? Switching tabs would switch the output as well.

### CODE REVIEW

File: src/components/Workspace/Workspace.tsx [hunk 1]
```diff
@@ -99,25 +126,35 @@ export default function Workspace(): JSX.Element {
...
+      setInputTab('judge');
+    } else {
+      setProblemData(undefined);
+      setJudgeResults(newJudgeResults);
     }
   };
   useEffect(() => {
     updateProblemData(settings.judgeUrl);
-  }, [settings?.judgeUrl]);
+  });
```
Comment (thecodingwizard) [inline src/components/Workspace/Workspace.tsx:155]: why remove `[settings?.judgeUrl]`? Won't doing this make the function run on every render (which seems bad for performance)?

File: src/pages/EditorPage.tsx [hunk 2]
```diff
@@ -150,23 +150,104 @@ export default function EditorPage(props: EditorPageProps): JSX.Element {
...
       }
-    )
+    );
+  };
+
+  const [inputTab, setInputTab] = useAtom(inputTabAtom);
+  const tabsList = useAtomValue(tabsListAtom);
+  const inputTabIndex = useAtomValue(inputTabIndexAtom);
+  useEffect(() => {
+    if (inputTabIndex === tabsList.length) {
```
Comment (thecodingwizard) [inline src/pages/EditorPage.tsx:160]: should this be `>=` instead of `===`?

File: src/pages/EditorPage.tsx [hunk 3]
```diff
@@ -150,23 +150,104 @@ export default function EditorPage(props: EditorPageProps): JSX.Element {
...
+
+  const cleanAndReplace = (
+    output: string
+  ): { replaced: string; cleaned: string } => {
+    const replaced = output.replace(/ /g, '\u2423'); // spaces made visible
+    const lines = output.split('\n');
+    for (let i = 0; i < lines.length; ++i) lines[i] = lines[i].trim();
+    const cleaned = lines.join('\n').trim(); // remove leading / trailing whitespace on each line, trim
+    return { replaced, cleaned };
+  };
```
Comment (thecodingwizard) [inline src/pages/EditorPage.tsx:189]: this should probably be moved to a utility file :O

File: src/pages/EditorPage.tsx [hunk 4]
```diff
@@ -150,23 +150,104 @@ export default function EditorPage(props: EditorPageProps): JSX.Element {
...
+          data.stdout = replaced; // show the extra whitespace
+        } else {
+          data.status.description = 'Wrong Answer';
+        }
+      }
+    }
+    if (prefix && !data.status.description.includes('Compilation'))
+      // on compilation error
+      data.status.description = prefix + data.status.description;
+  };
```
Comment (thecodingwizard) [inline src/pages/EditorPage.tsx:218]: same with this (move to utility file)

**Review by thecodingwizard — APPROVED** [review]
nice, looks good!! (except for some minor things, see above comments)

File: src/pages/EditorPage.tsx (hunk 2 above)
Comment (bqi343) [reply src/pages/EditorPage.tsx:163]: I don't think that ever happens though? I assumed that when `tabsList` is recomputed, `inputTabIndex` is also recomputed, and it's always in the range `[0, tabsList.length]`.

Comment (thecodingwizard) [reply src/pages/EditorPage.tsx:150]: oh, makes sense 👍

**Review by thecodingwizard — APPROVED** [review]
looks good!

**Pull request merged**